### ISSUE TITLE
Visualize 3D Smoke and Fire Output for EVACUATION_MC_MODE=.TRUE.

### ISSUE THREAD

**Issue by godisreal** [firemodels/fds#7115]
Is it possible to show 3D smoke in smokeview when EVACUATION_MC_MODE=.TRUE.?  
There are several evacuation runs for the same fire/smoke scenario, and thus I set EVACUATION_MC_MODE=.TRUE. so that FED file is directly read in, not recalculated, but it turns out that 3D smoke cannot be shown in smokeview.  There is no load/unload->3Dsmoke. Thanks a lot.

**Comment by tkorhon1** [comment]
Yes, there is. See the FDS-SMV discussion forum thread "FDS+EVAC: Smokeview visualization".

TimoK

**Comment by godisreal** [comment]
Hi, Timo.  
Many thanks.  Your guidance is helpful.  So the key idea is using old .fed file to generate new .prt5 in EVACUATION_MC_mode, and then replace the old .prt5 by the new one.  
So .fed file includes all the fire/smoke data needed by the evacuation module, and .s3d file is only used to show 3D smoke.  Is that right?  .s3d file is not directly related to the evac computation.  
Best Regards, 
Peng WANG

**Comment by tkorhon1** [comment]
Yes, about so. But you should be carefull, when copying the new .prt5 file to the old ones. The mesh numbers have changed. In the original calculation the mesh number = fire meshes + evacuation meshes. And in the EVACUATION_MC_MODE=.TRUE. case there are no fire meshes, so you should keep this in mind, when you copy the evacuation mesh .prt5 files.

It might be easier to edit the CHID.smv text file, where the file names are given. In this case,
it is nice to change the CHID string of the MC mode run, so you can have the new MC mode
run files in the same directory as the old ones and the .smv file that you edit. I am not sure,
if you can give names that include path in the .smv file. You could try this.

Or just do the copying, but do a DOS bat file (or shell script in Unix/Linux) that copies the
files, so you do not need to do much work with mouse.

Timo

**Comment by godisreal** [comment]
Hi, Timo. What is a feature request Issue? I started this issue previously, and you may reopen it if you would like to work on this problem.  So I copy and paste your comments to here.  

* add _evac_ to all evacuation output file names. This makes it easier to keep the old fire results.
   (One is not accidentally overwriting fire results.)

 * Add a keyword to MISC that controls the "mesh numbering" of the evacuation prt5 files (and
   possibly also the evacuation=true SLCF files for the guiding flow fields for evacuation): Make
   the evacuation "mesh numbers" in the prt5 file names be same as in the fire+evacuation
   calculation (I.e., one shifts the MC mode =true evacuation mesh number by the number
   of the fire meshes)

* Add a keyword (to MISC?) that puts a "ordinal number" to the evacuation .prt5 files (and
  to chid_evac.out and chid_evac.csv files also?) so that different MC_mode=true runs
  do not overwrite "previous MC mode results". 

* Check, which files the present MC_mode=true calculation writes. Which of these should
   not be written (i.e., do not contain any good information, like chid_hrr.csv). Note, that
   the chid.out file will be (re)written always. If the fire calculation chid.out file is lost and
   one wants it back, this is quite easy. Run just a second (or T_BEGIN=0), to get the
   useful information that is printed, when FDS reads the inputs (like material properties,
   pyrolysis parameters, reaction parameters, species informtation etc.).

**Comment by tkorhon1** [comment]
Ok, thanks! I'll reopen this Issue. Might be that "a feature reques Issue" was something that you could do in the Google Code Issue tracker. I do not know how things are here at GitHub. But this issue is good for this current purpose. They are the same thing. I.e., how to show smoke and agents at the same time in Smokeview.

TimoK

**Comment by godisreal** [comment]
Is it possible to avoid replacing chid_hrr.csv when BTEST(I_EVAC, 4) is false?
I think I just need to add a judgement somewhere either in dump.f90 or main.f90...
Thanks in advance.  
@mcgratta 

```
IF (APPEND) THEN
   OPEN(LU_HRR,FILE=FN_HRR,FORM='FORMATTED',STATUS='OLD')
   CALL APPEND_FILE(LU_HRR,2,T)
ELSE
   OPEN(LU_HRR,FILE=FN_HRR,FORM='FORMATTED',STATUS='REPLACE')
   WRITE(TCFORM,'(A,I4.4,A)') "(",12+N_ZONE_TMP,"(A,','),A)"
   WRITE(LU_HRR,TCFORM) 's','kW','kW','kW','kW','kW','kW','kW','kW','kW','kW','kg/s','kg/s',('Pa',N=1,N_ZONE_TMP)
   IF (N_ZONE_TMP>0) THEN
      WRITE(LU_HRR,TCFORM) 'Time','HRR','Q_RADI','Q_CONV','Q_COND','Q_DIFF','Q_PRES','Q_PART','Q_GEOM','Q_ENTH','Q_TOTAL',&
                           'MLR_FUEL','MLR_TOTAL',(TRIM(P_ZONE(N)%ID),N=1,N_ZONE_TMP)
   ELSE
      WRITE(LU_HRR,TCFORM) 'Time','HRR','Q_RADI','Q_CONV','Q_COND','Q_DIFF','Q_PRES','Q_PART','Q_GEOM','Q_ENTH','Q_TOTAL',&
                           'MLR_FUEL','MLR_TOTAL'
   ENDIF
ENDIF
```

**Comment by godisreal** [comment]
The program has some stop problem after FN_PART is changed.

**Comment by tkorhon1** [comment]
I just commited a new source code. It has now "evacuation bug fixes" for the changes:

1) Glenn modified the CHID_prt5.bnd file format. And this made the fire+evacuation part to crash, at least in the debug compilation.

2) Kevin changed the way the pressure zones are initialized/read in/defined. I just fixed this, hopefully. My simple test case seems to run, but I have no checked that the evacuation results are reasonable for my other V&V cases. 

I_EVAC = 4 (MC mode): 

dump.f90 SUBROUTINE DUMP_HRR(T,DT)
Add here:

ID (I_EVAC==4) RETURN

dump.f90 line 797 and onwards: Put here line like "IF(.NOT. I_EVAC==4) THEN .... END IF.
So, do the things that there are just when i_evac is not 4.

Or might also work here, dump.f90 line 297:
FN_HRR   = TRIM(CHID)//'_hrr.csv' ==> FN_HRR   = TRIM(CHID)//'_mc_evac_hrr.csv'
if I_EVAC==4.

This later one might be the easiest thing to do. Actually, similar thing for other files that should not be overwritten could be done. Actually, one could rename all evacuation MC mode files differently. I should check, if this would be easy by renaming the CHID in the read.f90 routine for I_EVAC==4 mode. One could rename CHID by adding an ordial number like CHID => CHID_1. If this is already found on the current directory, try CHID_2, and so on. This way the MC mode would not overwrite the information of the old cases. Well, sometimes you want to overwrite, so some user input might be added for this. There is already OVERWRITE input on the &MISC namelist. I could see, if this could also be used for evacuation MC mode or not (evacuation MC mode does not do any fire calculation, i.e., no fire outputs so OVERWRITE could be then used for evacuation).

**Comment by godisreal** [comment]
I use the following FN_PART (dump.f90) so that I can easily reuse .smv to show 3D smoke when EVACUATION_MC_MODE=.TRUE.  However, the program cannot stop as normal.  When the computation is complete, I have to use CTRL+C to terminate the program.  Does anyone know how to solve the problem?  Thanks.  

@gjaeger

Below is my revision in dump.f90

```

Count_FIRE_MESH_LOOP: DO NM=1,NMESHES
	IF (EVACUATION_ONLY(NM)) THEN
		CYCLE Count_FIRE_MESH_LOOP
	ELSE
		NFM = NFM+1
	ENDIF
ENDDO Count_FIRE_MESH_LOOP		
```

```

   ! Particle Files for Evacuation

   IF (EVACUATION_ONLY(NM)) THEN
      IF (EMESH_INDEX(NM)>0) THEN
         LU_PART(NM) = GET_FILE_NUMBER()
         IF (NMESHES>1) THEN
            WRITE(FN_PART(NM),'(A,I4.4,A)') TRIM(CHID)//'_evac_',NM-NFM,'.prt5'
         ELSE
            WRITE(FN_PART(NM),'(A,A)') TRIM(CHID),'_evac.prt5'
         ENDIF
      END IF
   ENDIF
```

**Comment by tkorhon1** [comment]
Well, above might work, but one could change it a little bit:

SUBROUTINE ASSIGN_FILE_NAMES

! Assign names and logical units for all output files

USE COMP_FUNCTIONS
USE EVAC, ONLY : EMESH_INDEX
INTEGER :: NM,I,N, NFM
CHARACTER(LABEL_LENGTH) :: CFORM

clip-clip-clip

   ! Particle Files for Evacuation

NFM = 0
Count_FIRE_MESH_LOOP: DO NM=1,NMESHES
	IF (EVACUATION_ONLY(NM)) CYCLE Count_FIRE_MESH_LOOP
	NFM = NFM+1
ENDDO Count_FIRE_MESH_LOOP	

   IF (EVACUATION_ONLY(NM)) THEN
      IF (EMESH_INDEX(NM)>0) THEN
         LU_PART(NM) = GET_FILE_NUMBER()
         LU_PART(NM+NMESHES) = GET_FILE_NUMBER()
         WRITE(FN_PART(NM),'(A,I4.4,A)') TRIM(CHID)//'_evac_',NM-NFM,'.prt5'
         WRITE(FN_PART(NM+NMESHES),'(A,I4.4,A)') TRIM(CHID)//'_evac_',NM-NFM,'.prt5.bnd'
      ENDIF
   ENDIF

This is based on the latest GitHub source code version of dump.f90.

This format might be nice, so the evacuation output files are numbered using 
evacuation mesh numbers.

I added the above on my dump.f90. I'll test it and if it works, I will commit it to Github at some point. I need to check the FDS+Evac Guide, if I should write something to it. And find some old fds-smv discussion forum threads that deals how to show the mc_mode=true results.

TimoK

**Comment by gjaeger** [comment]
I tried out the modifications with the example /Verification/Evacuation/evac_memory_test1.fds.  For this, I compiled FDS (mpi_gnu_osx_64) with commit fd77834e0.

```
 Revision         : FDS6.7.3-388-gfd77834e0-master
 Revision Date    : Sun Jan 19 16:28:57 2020 -0500
 Compiler         : gfortran
 Compilation Date : Jan 20, 2020  10:05:56

 MPI Enabled;    Number of MPI Processes:       1
 OpenMP Disabled

 MPI version: 3.1
 MPI library version: Open MPI v4.0.2, package: Open MPI brew@Catalina.local Distribution, ident: 4.0.2, repo rev: v4.0.2, Oct 07, 2019
```

I took the following steps:
1. run with one fire & one evacuation mesh
2. run with one evacuation mesh (no fire mesh)

_**step 1**_

_evac_memory_test1.fds_

```
&MISC EVACUATION_MC_MODE=.FALSE., 
      NO_EVACUATION=.FALSE., 
      EVACUATION_DRILL=.FALSE. /
...
&SLCF PBZ= 1.5, QUANTITY='VELOCITY', VECTOR=.TRUE., EVACUATION=.TRUE. /
&SLCF PBZ= 1.500, QUANTITY='VISIBILITY' /
```
This is an extract of the smw-file: 
```
...
SMOKF3D     1
 evac_memory_test1_0001_01.s3d
 SOOT DENSITY
 rho_C0.9H0.1
 kg/m3
SMOKF3D     1
 evac_memory_test1_0001_02.s3d
 HRRPUV
 hrrpuv
 kW/m3
SMOKG3D     1
 evac_memory_test1_0001_03.s3d
 TEMPERATURE
 temp
 C
SMOKG3D     1
 evac_memory_test1_0001_04.s3d
 CARBON DIOXIDE DENSITY
 rho_CO2
 kg/m3
SLCF     1 # STRUCTURED &     0   200     0   160     6     6 !      2
 evac_memory_test1_0001_01.sf
 SOOT VISIBILITY
 VIS_C0.9H0.1
 m
SLCF     2 # STRUCTURED &     0   200     0   160     1     1 !      1
 evac_memory_test1_0002_01.sf
 VELOCITY
 vel
 m/s
SLCF     2 # STRUCTURED &     0   200     0   160     1     1 !      1
 evac_memory_test1_0002_02.sf
 U-VELOCITY
 U-VEL
 m/s
SLCF     2 # STRUCTURED &     0   200     0   160     1     1 !      1
 evac_memory_test1_0002_03.sf
 V-VELOCITY
 V-VEL
 m/s
SLCF     2 # STRUCTURED &     0   200     0   160     1     1 !      1
 evac_memory_test1_0002_04.sf
 W-VELOCITY
 W-VEL
 m/s
EVA5     2    1.2000
 evac_memory_test1_evac_0001.prt5
  1
  1
```

The smv-file includes all files for `SMOKE3D` and  `SLCF`.

_**step 2**_

_evac_memory_test1.fds_ with modification

```
&MISC EVACUATION_MC_MODE=.TRUE., 
      NO_EVACUATION=.FALSE., 
      EVACUATION_DRILL=.FALSE. /
...
&SLCF PBZ= 1.5, QUANTITY='VELOCITY', VECTOR=.TRUE., EVACUATION=.TRUE. /
&SLCF PBZ= 1.500, QUANTITY='VISIBILITY' /
```

This is an extract of the smw-file: 
```
... 
SLCF     1 # STRUCTURED &     0   200     0   160     1     1 !      1
 evac_memory_test1_0001_01.sf
 VELOCITY
 vel
 m/s
SLCF     1 # STRUCTURED &     0   200     0   160     1     1 !      1
 evac_memory_test1_0001_02.sf
 U-VELOCITY
 U-VEL
 m/s
SLCF     1 # STRUCTURED &     0   200     0   160     1     1 !      1
 evac_memory_test1_0001_03.sf
 V-VELOCITY
 V-VEL
 m/s
SLCF     1 # STRUCTURED &     0   200     0   160     1     1 !      1
 evac_memory_test1_0001_04.sf
 W-VELOCITY
 W-VEL
 m/s
EVA5     1    1.2000
 evac_memory_test1_evac_0001.prt5
  1
  1
```

The SLCF files belong to the command: 
`&SLCF PBZ= 1.5, QUANTITY='VELOCITY', VECTOR=.TRUE., EVACUATION=.TRUE. /`

The output file numbering and naming works. This problem is solved.

There is still a problem: The references to the `SMOKE3D` and  `SLCF` of the fire mesh are overwritten.

@tkorhon1 , @gforney: 
Shouldn't we add a loop in `dump.f90` (like [line 509 ff](https://github.com/firemodels/fds/blob/fd77834e029b4dc30affa4d03fb0913498add341/Source/dump.f90#L509)) that checks if there is an s3d file in the folder? If it exists, then it should be placed in the SMV-file. Does that make sense? Do you have any suggestions?

**Comment by godisreal** [comment]
Hi, Gregor, 

Thank you for your test.  
Your logic is correct.  Another method is to judge if there is already a chid.smv generated in the folder.  If so, then judge if it is to be replaced or not.  Timo suggested an easy solution by 
revising FN_SMV based on I_EVAC, and it works roughly fine, but I recently find a small problem for simulation without evacuation.  Maybe you can find a smarter way to solve the problem, but it requires more changes of the code.  

I attache a part of code in dump.f90.  Maybe it is helpful to implement your idea.  You are welcome to contribute further.  

```
IF (.NOT.OVERWRITE) THEN
   INQUIRE(FILE=FN_OUTPUT,EXIST=EX)
   IF (EX) THEN
      WRITE(LU_ERR,'(A,A,A)')  'ERROR: OVERWRITE=.FALSE. and the file ',TRIM(FN_OUTPUT),' exists.'
      STOP
   ENDIF
ENDIF
```

Hi, Timo.  I run a test case without evacuation, and it shows that I_EVAC<16.  Is it I_EVAC=0 by default for an example with no evacuation?  What is a good method to judge if there is only evaucaiton or only fire simulaiton?  
I current use IF (I_EVAC<16 .AND. ANY(EVACUATION_ONLY)) THEN...
Your guidance is much appreciated.

**Comment by gjaeger** [comment]
> There is already OVERWRITE input on the &MISC namelist. I could see, if this could also be used for evacuation MC mode or not (evacuation MC mode does not do any fire calculation, i.e., no fire outputs so OVERWRITE could be then used for evacuation).

@tkorhon1:  I tried out your idea with

```
&MISC EVACUATION_MC_MODE=.TRUE.,
      NO_EVACUATION=.FALSE.,
      EVACUATION_DRILL=.FALSE.
      OVERWRITE=.FALSE. /
```
I obtain the following result:

```zsh
% fds_gnu_osx_64 evac_memory_test1.fds

 Starting FDS ...

 MPI Process      0 started on zam433

 Reading FDS input file ...

 EVAC: Emesh 1 FF1stFloor has 4 door flow fields
 FDS+Evac pressure method : FFT
ERROR: OVERWRITE=.FALSE. and the file evac_memory_test1.out exists.
```
Without modification the function `OVERWRITE` is not helpful.

Any suggestions?

_supplement_:

the if-statement in [dump.f90](https://github.com/firemodels/fds/blob/7e61b3cc39f407c52300b15957df148a0ef4518e/Source/dump.f90#L273) quoted by Peng only checks the existence of the diagnostic output file.

```f90
! Diagnostic Output File

FN_OUTPUT = TRIM(CHID)//'.out'
IF (.NOT.OVERWRITE) THEN
   INQUIRE(FILE=FN_OUTPUT,EXIST=EX)
   IF (EX) THEN
      WRITE(LU_ERR,'(A,A,A)')  'ERROR: OVERWRITE=.FALSE. and the file ',TRIM(FN_OUTPUT),' exists.'
      STOP
   ENDIF
ENDIF
```

**Comment by godisreal** [comment]
Hi, Gregor.  Thank you for the information.  I read FDS manual about OVERWRITE in MISC.  Yes, it is useful to some extent.  If there is already fire data generated in last run, OVERWRITE=.FALSE. will stop the program so that users can move the data to somewhere else.  Here the key issue is .smv file.  As you suggested, there should be a smarter way to avoid overwriting .smv in certain situation.  The logic is given as below. 
```

IF (no .smv in the folder)  --> Generate .smv file
ELSE 
   IF (Only Evac Simulation) --> Do not overwrite .smv file
   ELSE --> Overwrite .smv file 
```

Hi, Timo,
Considering Gregor's suggestion I attach some code in SUBROUTINE WRITE_SMOKEVIEW_FILE as below.  Maybe we can realize a smarter logic in the judgement sentence.  (Perhaps adding I_EVAC in the judgement)  

```
! Open up the Smokeview ".smv" file

INQUIRE(FILE=FN_SMV,EXIST=EX)
IF (.NOT.EX .AND. APPEND) THEN
   WRITE(MESSAGE,'(A,A,A)') "ERROR: The file, ",TRIM(FN_SMV),", does not exist. Set RESTART=.FALSE."
   CALL SHUTDOWN(MESSAGE)
ENDIF

IF (APPEND) THEN
   OPEN(LU_SMV,FILE=FN_SMV,FORM='FORMATTED', STATUS='OLD',POSITION='APPEND')
   RETURN
ENDIF

OPEN(LU_SMV,FILE=FN_SMV,FORM='FORMATTED',STATUS='REPLACE')

! Write out TITLE

WRITE(LU_SMV,'(A)') 'TITLE'
WRITE(LU_SMV,'(1X,A)')  TRIM(TITLE)
```

**Comment by gjaeger** [comment]
@godisreal 
I don't think it's good that parallel to the main repository a further development takes place. This is not good for the project.

Why don't you describe your proposed solution for `dump.f90` here? 

```f90
IF (EX .AND. I_EVAC<16 .AND. ANY(EVACUATION_ONLY)) THEN
    OPEN(LU_SMV,FILE=FN_SMV,FORM='FORMATTED', STATUS='OLD',POSITION='APPEND')
    RETURN
 ENDIF
```

You could ask Glenn, Randy or Kevin if  if they agree with your proposed solution. Then we could avoid sleeping pull requests.

**Comment by godisreal** [comment]
Hi, Gregor, 
If you read the code, you will know the key issue is the judgement sentence, i.e., I_EVAC<16 .AND. ANY(EVACUATION_ONLY), and Timo is the most qualified to review it and merge this change. Glenn proabably is not interested because the key idea refers to evac code, not smokeview code.  I suppose you do not like Fortran, do you?  So if you read and understand the code, you will know who is in charge of this change.  
In addition I give PR to Timo's repo because I just modifed the code last week and tested it quickly in one example.  I do not have much time to test it by many examples.  That's not my focus.  So if Timo can confirm this change, then it is done. Also, if you are interested, maybe you can help to do some testing work, and I can guide you to do that.  Unless I can confirm this change is correct or get some feedback from Timo, I will not give a PR here.  
Anyway, this is not an urgent change.  So there is no need to give much pressure on anyone.  Let Timo review it first.  If Timo does not respond, I will consider submitting a PR here as you suggested.  Many thanks.

**Comment by gjaeger** [comment]
@godisreal , you misunderstand me.

The more changes are made to the source code, the more difficult it is to merge the changes. A pull request to another fork repository makes development more difficult. Why not make the change in your repository? You can refer to it here. If your change is approved, you can suggest a pull request. All other ways are not helpful in my opinion.

I named Glenn, Randy and Kevin because the change has nothing directly to do with the FDS+Evac model. In my opinion it is easier to look into your fork repository. There it is easier to try out changes instead of stressing the main repository unnecessarily. Overall, your procedure does not correspond to the usual [workflow](https://github.com/firemodels/fds/blob/master/CONTRIBUTING.md) and the [developer commit guidelines](https://github.com/firemodels/fds/wiki/Developer-Commit-Guidelines).

Back to your modification:
Which cases would you want to check? What do you mean by _many examples_? Can you describe cases where your modification should work? Then we can create the required verification tests. This will increase acceptance.

**Comment by rmcdermo** [comment]
I agree with Gregor.  If evac.f90 is to be maintained, code modifications should be considered, by Timo or Glenn or whomever, for the central repo.

**Comment by godisreal** [comment]
OK.  Since you both support direct PR to central repo, I will plan to submit a draft PR here.  Maybe Gregor can help to run more cases, and I will let him know how he can help.  It is better to have several testing cases before finalize the code change.  
Timo also suggested another easy method to solve the problem, and I have been using that method for a while.  Now I also need to compare the two ways.  
Anyway, this is not an urgent issue, and there is no need to push anyone.  It is better to complete the task step by step.  

Let me present two solutions to the problem as below.  There are two method as Timo suggested.  

Solution 1: WRITE_SMOKEVIEW_FILE
```
IF (EX .AND. I_EVAC<16 .AND. ANY(EVACUATION_ONLY)) THEN
   OPEN(LU_SMV,FILE=FN_SMV,FORM='FORMATTED', STATUS='OLD',POSITION='APPEND')
   RETURN
ENDIF
```

Solution 2: ASSIGN_FILE_NAMES
```
IF (I_EVAC<16 .AND. ANY(EVACUATION_ONLY)) THEN
	FN_SMV = TRIM(CHID)//'_evac.smv'
ELSE
	FN_SMV = TRIM(CHID)//'.smv'
ENDIF
```

Now I campare two methods as below.  
Solution 1 generates only one .smv file.  This method is user-friendly, but I am not sure if it is OK for fire simulation.  So more tests are needed.
Solution 2 generates two .smv file.  This method is less smarter than Solution 1, but it may be safer and more reliable in current situation because it rewrites .smv as another file in MC_Mode.  

The testing work is explained as below.  

Run any example with both fire and evacuation for three times and check how .smv file is changed and observe if 3D smoke is shown properly in smokeview.  
```
1.  MISC EVACUATION_MC_MODE=.TRUE. EVACUATION_DRILL=.TRUE.  --->  No 3D Smoke; .smv file is created
2.  MISC EVACUATION_MC_MODE=.FALSE. EVACUATION_DRILL=.FALSE.  --->  3D Smoke; .smv file is updated
3.  MISC EVACUATION_MC_MODE=.TRUE. EVACUATION_DRILL=.FALSE.  --->   3D Smoke; .smv file is not changed
```

The above solutions enable smokeview to visualize hazard-and-human interaction in multiple evacuation run (MC_MODE), and thus provide an easy tool to solve some other issues.  If Gregor is interested, please feel free to test any solutions and let me know if there is any problem.  Thanks.

**Comment by godisreal** [comment]
Let me present two solutions to the problem as below.  There are two method as Timo suggested.  

Solution 1: WRITE_SMOKEVIEW_FILE
```
IF (EX .AND. I_EVAC<16 .AND. ANY(EVACUATION_ONLY)) THEN
   OPEN(LU_SMV,FILE=FN_SMV,FORM='FORMATTED', STATUS='OLD',POSITION='APPEND')
   RETURN
ENDIF
```

Solution 2: ASSIGN_FILE_NAMES
```
IF (I_EVAC<16 .AND. ANY(EVACUATION_ONLY)) THEN
	FN_SMV = TRIM(CHID)//'_evac.smv'
ELSE
	FN_SMV = TRIM(CHID)//'.smv'
ENDIF
```

Now I campare two methods as below.  
Solution 1 generates only one .smv file.  This method is user-friendly, but I am not sure if it is OK for fire simulation.  So more tests are needed.
Solution 2 generates two .smv file.  This method is less smarter than Solution 1, but it may be safer and more reliable in current situation because it rewrites .smv as another file in MC_Mode.  

The testing work is explained as below.  

Run any example with both fire and evacuation for three times and check how .smv file is changed and observe if 3D smoke is shown properly in smokeview.  
```
1.  MISC EVACUATION_MC_MODE=.TRUE. EVACUATION_DRILL=.TRUE.  --->  No 3D Smoke; .smv file is created
2.  MISC EVACUATION_MC_MODE=.FALSE. EVACUATION_DRILL=.FALSE.  --->  3D Smoke; .smv file is updated
3.  MISC EVACUATION_MC_MODE=.TRUE. EVACUATION_DRILL=.FALSE.  --->   3D Smoke; .smv file is not changed
```

The above solutions enable smokeview to visualize hazard-and-human interaction in multiple evacuation run (MC_MODE), and thus provide an easy tool to solve some other issues.  If Gregor is interested, please feel free to test any solutions and let me know if there is any problem.  Thanks.

**Comment by gjaeger** [comment]
@godisreal I'll look at it.

**Comment by tkorhon1** [comment]
Hi! I have been quite busy (family reasons), so I have not followed this issue lately. Is there something for me to do? I should update the manual, if this Issue has already found the ultimate solution to the problem. If you are still looking the best way of doing things, then I'll try to read your comments above and see, where things are right now.

I'll might have some time to check the FDS+Evac related things in the coming weeks. I have now survived the start of this year. But things might change rapidly due to the corona stuff. If they close daycare here, then I'm out of the business until they open those after the corona crisis.

TimoK

**Comment by rmcdermo** [comment]
@tkorhon1  While you have time, it would be good to have at least an online chat with both @gjaeger and @godisreal  to get organized for long-term FDS+Evac maintenance.  @gjaeger has generously offered to help with issues and documentation.

**Comment by tkorhon1** [comment]
Well, I am able to chat Mon-Fri 09:30 - 14:30 UTC+2 (+daylight saving time, if appropriate, summer time begins 29th April => UTC+3). The other time window for me would be 22:30-24:00 UTC+2 (Sun-Sat). 

Well, "able to chat"? I am not in any social media. 

TimoK

**Comment by gjaeger** [comment]
@tkorhon1 
This issue contains a few suggested solutions. We should talk about the ideas and then implement a version in the code.

**Comment by tkorhon1** [comment]
Well, I should have done this already, but I have been too busy at home and at office. I'm now having a couple of hours to check FDS+Evac related issues, then I'm on holiday, and in August back at home office (we still continue to word at home even the corona situation in Finland is really good at this time). In August I should have some time to do FDS+Evac stuff, because I have just finnished four customer assigments that I had to do before summer holidays. So, after the holidays, there is no deadline straight ahead.
